erratic test behaviour

im interested in making this code compatible with python 2 and 3, python 3 works pretty well using 2to3.py, but the get_root() function isn't working for me.

Im using windows 7, testing is done on unaltered code from this repo

***running python2 -m py.test tests***
collected 33 items

during testing i get a few uac dialog boxes which i confirm

only ever passed once without errors,  There are sometimes 2 sometimes 3 tests that are failed.

***testing with tox -e py27:***
collected 30

No uac dialog prompts

The tests pass 50% of the time under tox, when it fails its because of the error pasted below.


I'd like some guidance on my next actions, i feel i need to get the tests working 100% of the time before moving on. Not sure what the best course of action is.

Does any one know why tox is collecting different amount of tests?

**mytox.ini**
    [tox]
    envlist = py27, py33, py34
    [testenv]
    deps=pytest 
    commands=py.test


***Here is the error that gets thrown the most when running under tox and py.test***
```python
self = <esky.tests.test_esky.TestPatch_pybsdiff testMethod=test_diffing_back_and
_forth>

    def tearDown(self):
        esky.patch.bsdiff4 = self.__orig_bsdiff4
>       return super(TestPatch_pybsdiff,self).tearDown()

test_esky.py:765:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
test_esky.py:649: in tearDown
    shutil.rmtree(self.workdir)
C:\Python27\lib\shutil.py:247: in rmtree
    rmtree(fullname, ignore_errors, onerror)
C:\Python27\lib\shutil.py:256: in rmtree
    onerror(os.rmdir, path, sys.exc_info())
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

path = 'D:\\STORAGE\\Dropbox\\programming\\testesky\\unchanged\\esky\\tests\\pat
ch-test-temp.win32\\source'
ignore_errors = False, onerror = <function onerror at 0x03BBFA70>

    def rmtree(path, ignore_errors=False, onerror=None):
        """Recursively delete a directory tree.

        If ignore_errors is set, errors are ignored; otherwise, if onerror
        is set, it is called to handle the error with arguments (func,
        path, exc_info) where func is os.listdir, os.remove, or os.rmdir;
        path is the argument to that function that caused it to fail; and
        exc_info is a tuple returned by sys.exc_info().  If ignore_errors
        is false and onerror is None, an exception is raised.

        """
        if ignore_errors:
            def onerror(*args):
                pass
        elif onerror is None:
            def onerror(*args):
                raise
        try:
            if os.path.islink(path):
                # symlinks to directories are forbidden, see bug #1669
                raise OSError("Cannot call rmtree on a symbolic link")
        except OSError:
            onerror(os.path.islink, path, sys.exc_info())
            # can't continue even if onerror hook returns
            return
        names = []
        try:
            names = os.listdir(path)
        except os.error, err:
            onerror(os.listdir, path, sys.exc_info())
        for name in names:
            fullname = os.path.join(path, name)
            try:
                mode = os.lstat(fullname).st_mode
            except os.error:
                mode = 0
            if stat.S_ISDIR(mode):
                rmtree(fullname, ignore_errors, onerror)
            else:
                try:
                    os.remove(fullname)
                except os.error, err:
                    onerror(os.remove, fullname, sys.exc_info())
        try:
>           os.rmdir(path)
E           WindowsError: [Error 145] The directory is not empty: 'D:\\STORAGE\\
Dropbox\\programming\\testesky\\unchanged\\esky\\tests\\patch-test-temp.win32\\s
ource'

C:\Python27\lib\shutil.py:254: WindowsError
==================== 3 failed, 30 passed in 143.72 seconds ====================
Press any key to continue . . .
```

--- a/esky/tests/test_esky.py
+++ b/esky/tests/test_esky.py
@@ -244,11 +244,11 @@ class TestEsky(unittest.TestCase):
         deep_extract_zipfile(os.path.join(tdir,"dist","eskytester-0.1.%s.zip"%(platform,)),uzdir)
         with open(os.path.join(tdir,"dist","eskytester-0.3.%s.from-0.1.patch"%(platform,)),"rb") as f:
             esky.patch.apply_patch(uzdir,f)
-        shutil.rmtree(uzdir)
+        really_rmtree(uzdir)
         deep_extract_zipfile(os.path.join(tdir,"dist","eskytester-0.2.%s.zip"%(platform,)),uzdir)
         with open(os.path.join(tdir,"dist","eskytester-0.3.%s.from-0.2.patch"%(platform,)),"rb") as f:
             esky.patch.apply_patch(uzdir,f)
-        shutil.rmtree(uzdir)
+        really_rmtree(uzdir)
         #  Serve the updates at http://localhost:8000/dist/
         print "running local update server"
         server = HTTPServer(("localhost",8000),SimpleHTTPRequestHandler)
@@ -341,7 +341,7 @@ class TestEsky(unittest.TestCase):
         assert len(errors) == 1
         assert isinstance(errors[0],esky.EskyLockedError)
     finally:
-        shutil.rmtree(appdir)
+        really_rmtree(appdir)
 
  
   def test_esky_lock_breaking(self):
@@ -386,7 +386,7 @@ class TestEsky(unittest.TestCase):
         t2.join()
         assert len(errors) == 0, str(errors)
     finally:
-        shutil.rmtree(appdir)
+        really_rmtree(appdir)
 
 
   def test_README(self):
@@ -417,7 +417,7 @@ class TestFSTransact(unittest.TestCase):
         self.testdir = tempfile.mkdtemp()
 
     def tearDown(self):
-        shutil.rmtree(self.testdir)
+        really_rmtree(self.testdir)
 
     def path(self,path):
         return os.path.join(self.testdir,path)
@@ -646,7 +646,7 @@ class TestPatch(unittest.TestCase):
                     f.write(data)
 
     def tearDown(self):
-        shutil.rmtree(self.workdir)
+        really_rmtree(self.workdir)
 
     def test_patch_bigfile(self):
         tdir = tempfile.mkdtemp()
@@ -668,7 +668,7 @@ class TestPatch(unittest.TestCase):
             dgst3 = esky.patch.calculate_digest(os.path.join(tdir,"source"))
             self.assertEquals(dgst1,dgst3)
         finally:
-            shutil.rmtree(tdir)
+            really_rmtree(tdir)
 
     def test_diffing_back_and_forth(self):
         for (tf1,_) in self._TEST_FILES:
@@ -790,5 +790,5 @@ class TestFilesDiffer(unittest.TestCase):
         assert self._differs("onethreetwo","twothreeone",3,-2)
 
     def tearDown(self):
-        shutil.rmtree(self.tdir)
+        really_rmtree(self.tdir)
 

--- a/esky/util.py
+++ b/esky/util.py
@@ -499,7 +499,7 @@ def really_rmtree(path):
         if not os.path.exists(path):
             shutil.rmtree(path)
         #  This is a little retry loop that catches troublesome errors.
-        for _ in xrange(10):
+        for _ in xrange(100):
             try:
                 shutil.rmtree(path)
             except WindowsError, e:

